src/hash.cr

```diff
@@ -23,10 +24,36 @@ class Hash(K, V)
     @block = block
   end
 
+  # Creates a new empty `Hash` of *initial_capacity* with a *block* that handles missing keys.
+  #
+  # The *initial_capacity* is useful to avoid unnecessary reallocations
+  # of the internal buffer in case of growth. If you have an estimate
+  # of the maximum number of elements a hash will hold, the hash should
+  # be initialized with that capacity for improved performance.
+  #
+  # NOTE: *initial_capacity* defaults to 11 and an input of < 11 is ignored.
+  # NOTE: `#size` does not reflect capacity as in `Array`.
+  #
+  # ```
+  # new_hash = Hash(String, Int32).new(50) do |hash, key|
+  #   key.size
+  # end
+  # new_hash.size # => 0
+  # new_hash["zero"] = 0
+  # new_hash["two"] = 2
+  # new_hash["four"] # => 4
+  # ```
   def self.new(initial_capacity = nil, &block : (Hash(K, V), K -> V))
     new block, initial_capacity: initial_capacity
   end
 
+  # Creates a new empty `Hash` where the *default_value* is returned if key is missing.
+  #
+  # ```
```

Comment:
This could probably get a note that the *exact* same value will be used as default, which might have unwanted effects when it is a reference (because then all will point to the same instance).

Reply:
Ooh, I like that, good point. Will do on next commit once I hear a little feedback on the other comments here.